container-search/src/main/java/com/yahoo/search/dispatch/SearchPath.java

```diff
@@ -100,37 +102,46 @@ private SearchPath(List<NodeSelection> nodes, Integer group) {
     }
 
     private boolean isEmpty() {
-        return nodes.isEmpty() && group == null;
+        return nodes.isEmpty() && groups.isEmpty();
     }
 
-    private Group selectGroup(SearchCluster cluster) {
-        if (group != null) {
-            Optional<Group> specificGroup = cluster.group(group);
-            if (specificGroup.isPresent()) {
-                return specificGroup.get();
+    private Group selectRandomGroupWithSufficientCoverage(SearchCluster cluster, List<Integer> groupIds) {
+        for (int numRetries = 0; numRetries < groupIds.size()*2; numRetries++) {
+            int index = random.nextInt(groupIds.size());
+            int groupId = groupIds.get(index);
+            Optional<Group> group = cluster.group(groupId);
+            if (group.isPresent()) {
+                if (group.get().hasSufficientCoverage()) {
+                    return group.get();
+                }
             } else {
-                throw new InvalidSearchPathException("Invalid searchPath, cluster does not have " + (group + 1) + " groups");
+                throw new InvalidSearchPathException("Invalid searchPath, cluster does not have " + (groupId + 1) + " groups");
             }
         }
+        return cluster.groups().values().iterator().next();
+    }
 
-        // pick "anything": try to find the first working
-        ImmutableCollection<Group> groups = cluster.groups().values();
-        for (Group g : groups) {
-            if (g.hasSufficientCoverage()) {
-                return g;
+    private Group selectGroup(SearchCluster cluster) {
+        if ( ! groups.isEmpty()) {
+            List<Integer> potentialGroups = new ArrayList<>();
+            for (NodeSelection groupSelection : groups) {
+                for (int group = groupSelection.from; group < groupSelection.to; group++) {
+                    potentialGroups.add(group);
+                }
             }
+            return selectRandomGroupWithSufficientCoverage(cluster, potentialGroups);
         }
 
-        // fallback: first
-        return groups.iterator().next();
+        // pick any working group
+        return selectRandomGroupWithSufficientCoverage(cluster, cluster.groups().keySet().asList());
     }
 
     private static SearchPath parseElement(String element) {
         Pair<String, String> nodesAndGroup = halveAt('/', element);
         List<NodeSelection> nodes = parseNodes(nodesAndGroup.getFirst());
-        Integer group = parseGroup(nodesAndGroup.getSecond());
+        List<NodeSelection> groups = parseNodes(nodesAndGroup.getSecond());
```

Comment:
It would be nice if `parseNodes` changed name to reflect its new usage :)

Reply:
parseSelection it is. NodeSelection renamed to Selection.